scvi/train/_trainingplans.py

```diff
@@ -853,3 +865,215 @@ def configure_optimizers(self):
+
+    Parameters
+    ----------
+    flax_module
+        An instance of :class:`~scvi.module.base.JaxBaseModuleClass`. This object
+        should have callable `model` and `guide` attributes or methods.
+    n_steps_kl_warmup
+        Number of training steps (minibatches) to scale weight on KL divergences from 0 to 1.
+        Only activated when `n_epochs_kl_warmup` is set to None.
+    n_epochs_kl_warmup
+        Number of epochs to scale weight on KL divergences from 0 to 1.
+        Overrides `n_steps_kl_warmup` when both are not `None`.
+    """
+
+    def __init__(
+        self,
+        train_module: JaxBaseModuleClass,
+        val_module: JaxBaseModuleClass,
+        n_steps_kl_warmup: Union[int, None] = None,
+        n_epochs_kl_warmup: Union[int, None] = 400,
+        use_gpu: bool = False,
+        optim_kwargs: Optional[dict] = None,
+        **loss_kwargs,
+    ):
+        super().__init__()
+        self.module = train_module
+        self.validation_module = val_module
+        self._n_obs_training = None
+        self.loss_kwargs = loss_kwargs
+        self.n_steps_kl_warmup = n_steps_kl_warmup
+        self.n_epochs_kl_warmup = n_epochs_kl_warmup
+
+        self.automatic_optimization = False
+
+        # automatic handling of kl weight
+        self._loss_args = signature(self.module.loss).parameters
+        if "kl_weight" in self._loss_args:
+            self.loss_kwargs.update({"kl_weight": self.kl_weight})
+        self._rngs = None
+        self.use_gpu = use_gpu
+        # gets set by the initialization callback
+        self.state = None
+
+        if optim_kwargs is None:
+            self.optim_kwargs = dict(learning_rate=1e-3, eps=0.01, weight_decay=1e-6)
+        else:
+            self.optim_kwargs = optim_kwargs
+
+    def set_train_state(self, params, batch_stats=None):
+        weight_decay = self.optim_kwargs.pop("weight_decay")
+        # replicates PyTorch Adam
+        optimizer = optax.chain(
+            optax.additive_weight_decay(weight_decay=weight_decay),
+            optax.adam(**self.optim_kwargs),
+        )
+        state = TrainState.create(
+            apply_fn=self.module.apply,
+            params=params,
+            tx=optimizer,
+            batch_stats=batch_stats,
+        )
+        self.state = state
+
+    @property
+    def key_fn(self):
+        # if key is generated on CPU, model params will be on CPU
+        # we have to pay the price of a JIT compilation though
+        if self.use_gpu is False:
+            key = jax.jit(lambda i: random.PRNGKey(i), backend="cpu")
+        else:
+            # dummy function
+            def key(i: int):
+                return random.PRNGKey(i)
+
+        return key
+
+    @property
+    def rngs(self) -> Dict[str, jnp.ndarray]:
+        return self._rngs
+
+    @rngs.setter
+    def rngs(self, rngs: Dict[str, jnp.ndarray]):
+        self._rngs = rngs
+
+    def set_rngs(self, keys: List[str]):
```

Comment:
I think we can operate with PRNGKeyArrays here instead of list of ndarrays https://github.com/google/jax/blob/b051be4a7a163502ab871cf16e5f749d49622042/jax/_src/prng.py#L99.

Seems pretty feature rich and the default typing for most functions. For example random.split takes an array and returns an array

Reply:
This hides having explicit PRNGKeys floating around in a dict.

Reply:
ignore comment. Moved in https://github.com/scverse/scvi-tools/pull/1552